src/ops/base.rs

```diff
@@ -1,39 +1,221 @@
+use serde_json;
+use snowflake::ProcessUniqueId;
+use buf_redux::BufWriter;
+use buf_redux::strategy::WhenFull;
 use std::collections::HashMap;
+use std::fs;
+use std::fs::{File, OpenOptions};
+use std::path::PathBuf;
+use std::time::{Duration, Instant};
+use time;
+
+// 4k buffered log.
+const LOG_BUFFER_CAPACITY: usize = 4 * 1024;
+
+// We store at most this many write records before flushing to disk.
+const BUFFERED_WRITES_CAPACITY: usize = 512;
+
+// We spend at least this many milliseconds without flushing write records to disk.
+const BUFFERED_WRITES_FLUSH_INTERVAL_MS: u64 = 1000;
 
 /// Base is used to represent the root nodes of the distributary data flow graph.
 ///
 /// These nodes perform no computation, and their job is merely to persist all received updates and
 /// forward them to interested downstream operators. A base node should only be sent updates of the
 /// type corresponding to the node's type.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Base {
+    buffered_writes: Option<Records>,
+    durability: BaseDurabilityLevel,
+    durable_log: Option<BufWriter<File, WhenFull>>,
+    durable_log_path: Option<PathBuf>,
+    global_address: Option<NodeAddress>,
+    last_flushed_at: Option<Instant>,
     primary_key: Option<Vec<usize>>,
+    should_delete_log_on_drop: bool,
+
+    // This id is unique within the same process.
+    //
+    // TODO(jmftrindade): Figure out the story here.  While ProcessUniqueId is guaranteed to be
+    // unique within the same process, the assignment of ids is not deterministic across multiple
+    // process runs. This is just a tuple of 2 monotonically increasing counters: the first is per
+    // process, and the second is "within" that process.
+    unique_id: ProcessUniqueId,
+
     us: Option<NodeAddress>,
 }
 
+/// Specifies the level of durability that this base node should offer. Stronger guarantees imply a
+/// reduced write performance.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum BaseDurabilityLevel {
+    /// No durability at all: records aren't written to a log. This is the default durability level
+    /// if none is specified.
+    None,
+    /// Buffered writes: records are accumulated in an in-memory buffer and occasionally flushed to
+    /// the durable log, which may itself buffer in the file system. Results in large batched
+    /// writes, but offers no durability guarantees on crashes.
+    Buffered,
+    /// Synchronous writes: forces every record to be written to disk before it is emitted further
+    /// into the data-flow graph. Strong guarantees (writes are never lost), but high performance
+    /// penalty.
+    SyncImmediately,
+}
+
 impl Base {
     /// Create a base node operator.
-    pub fn new(primary_key: Vec<usize>) -> Self {
+    pub fn new(primary_key: Vec<usize>, durability: BaseDurabilityLevel) -> Self {
```

Comment:
Can we instead add another method, `new_durable`, that takes two arguments, leaving `new` to just take one, and give no durability? Would also reduce the diff since all the tests don't have to change :p